// flexlayout-react/lib/model/Floating.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Floating = void 0;
var Rect_1 = require("../Rect");
var TabNode_1 = require("./TabNode");
var Floating = /** @class */ (function () {
    function Floating(_rect, _parentId, _tabIndex, _tab) {
        this._rect = _rect;
        this._parentId = _parentId;
        this._tabIndex = _tabIndex;
        this._tab = _tab;
    }
    Object.defineProperty(Floating.prototype, "rect", {
        get: function () {
            return this._rect;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Floating.prototype, "tabIndex", {
        get: function () {
            return this._tabIndex;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Floating.prototype, "parentId", {
        get: function () {
            return this._parentId;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Floating.prototype, "tab", {
        get: function () {
            return this._tab;
        },
        enumerable: false,
        configurable: true
    });
    /** @internal */
    Floating._fromJson = function (json, model) {
        var tab = new TabNode_1.TabNode(model, json.tab, true);
        return new Floating(new Rect_1.Rect(json.left, json.top, json.width, json.height), json.parentId, json.tabIndex, tab);
    };
    Floating.prototype.toJson = function () {
        var json = {
            left: this.rect.x,
            top: this.rect.y,
            width: this.rect.width,
            height: this.rect.height,
            parentId: this.parentId,
            tabIndex: this.tabIndex,
            tab: this.tab.toJson()
        };
        return json;
    };
    return Floating;
}());
exports.Floating = Floating;
